### ISSUE TITLE
template_file: template attrib accepts interpolated string, but will not render vars

### ISSUE THREAD

**Issue by dayglojesus** [hashicorp/terraform#16699]
### Terraform Version
0.9.6 to Latest (0.11.0)

### Terraform Configuration Files
```
variable "my_template" {
  default = "some_key: $${insert_here}"
}

data "template_file" "document" {
  template = "${var.my_template}"

  vars {
    insert_here = "some_value"
  }
}

output "info" {
  value = "${data.template_file.document.rendered}"
}
```

### Debug Output
https://gist.github.com/dayglojesus/96375c2daa6ecedc398b729b2a585bca

### Expected Behavior
The properly escaped `$${insert_here}` should be evaluated *inside* the `template_file` resource as a template variable (`${insert_here}`), not rendered as a literal string.

### Actual Behavior
Template is rendered successfully, but the vars in the template are not substituted. Instead, the escaped vars ($$insert_here) are interpreted literally.
```
data.template_file.document:
  id = ed48139b5a707a192bc659d2d7303b35a5a74635029b80e3a0ef800e293cd00d
  rendered = some_key: ${insert_here}
  template = some_key: $${insert_here}
  vars.% = 1
  vars.insert_here = some_value

Outputs:

info = some_key: ${insert_here}
```

### Steps to Reproduce
1. `terraform init`
2. `terraform apply`

**Comment by apparentlymart** [comment]
Hi @dayglojesus! Sorry for this confusing behavior.

I suspect that this is behaving strangely because interpolation expressions are not expected in `default`, and therefore the usual pass to deal with them (which would also deal with the `$${` escape sequence) doesn't apply here.

Our new configuration language parser, which we're working on integrating now, should address this by applying consistent processing to all strings regardless of context. In the mean time, I think this could be worked around by applying the default in a context where interpolation _is_ allowed, such as in a local value:

```hcl
variable "my_template" {
  default = ""
}

locals {
  default_template = "some_key: $${insert_here}"
  template = "${var.my_template != "" ? var.my_template : local.default_template}"
}

data "template_file" "document" {
  template = "${local.template}"

  vars {
    insert_here = "some_value"
  }
}

output "info" {
  value = "${data.template_file.document.rendered}"
}
```

Since interpolation _is_ allowed in the `default_template` local value, it should be processed as expected and have that `$${` escape transformed to `${` as you wanted here.

The improved configuration language parser is a pretty disruptive change, and so we're going to roll out it quite cautiously over the course of a few releases -- opt-in only at first, I expect -- but it should address this problem with no additional effort just as a side-effect of the fact that the new parser integrates both the structural parsing (the blocks and attributes) and the expression parsing (what's inside the `${ ... }` sequences) into a single parser, which will avoid strange edge-cases where certain attribute values are parsed differently than others.

**Comment by dayglojesus** [comment]
This is a pretty easy workaround. Thanks @apparentlymart -- much obliged.

**Comment by teamterraform** [comment]
Hi again!

The configuration language improvements that @apparentlymart mentioned have now been released in Terraform 0.12. The change for variable defaults is that they are now parsed the same way as any other expression in the language, but references to other objects and function calls are both prohibited.

That means that interpolation sequences are themselves valid and must be escaped if a literal `${` is desired:

```hcl
variable "my_template" {
  default = "some_key: $${insert_here}"
}
```

Although this is unlikely to be useful most normal cases, you can now use interpolation sequences and template control sequences in `default` as long as they only work with constant values, and don't refer to any objects or functions:

```hcl
variable "my_template" {
  default = "This is rather ${"pointless"}, but it now works for consistency"
}
```

We wouldn't recommend doing things like that without a strong reason, but it became possible just as a result of parsing `default` the same way as any other expression in the language.
